### PR TITLE
EnumType and DomainSize

### LINKED ISSUE #122: Name collision checks for XSTS, enum type rework
More checks could be introduced to ensure that variable names, custom enum type names, and enum literals are unique. Also introducing a type for enums should be considered (instead of converting them to ints), which would only support a limited set of expressions (for example the `>` operator makes no sense for enums).

### PR DESCRIPTION
**Pull request by RipplB**
Fixes #122 
And also introduces DomainSize: every type should now be able to tell the number of possible values it can have

### CODE REVIEW

**Comment by RipplB** [comment]
@mondokm @szdan97

File: subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java
Comment (mondokm) [inline subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java:1]: Please add 1-2 tests instead of this removed test: one checking that literals of the same type can't be equal, and one checking that literals of different types with the same name can't be equal.

File: subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java [hunk 1]
```diff
@@ -131,14 +131,16 @@ public static Collection<Object[]> data() {
                 {"src/test/resources/model/choices.xsts", "src/test/resources/property/choices.prop",
                         false, XstsConfigBuilder.Domain.EXPL_PRED_COMBINED},
 
-                {"src/test/resources/model/literals.xsts", "src/test/resources/property/literals.prop",
```
Comment (mondokm) [inline subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java:134]: Please add 1-2 tests instead of this removed test: one checking that literals of the same type can't be equal, and one checking that literals of different types with the same name can't be equal.

File: subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java
Comment (mondokm) [inline subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java:1]: This seems OK for now, but the whole XstsType thing will have to be removed completely sonn

Comment (RipplB) [reply subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java:1]: Don't we need it for the overriden #createBoundExpr() to contain all enum literals' disjunction?

File: subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java (hunk 1 above)
Comment (RipplB) [reply subprojects/xsts/xsts-analysis/src/test/java/hu/bme/mit/theta/xsts/analysis/XstsTest.java:134]: Done in e757a60

File: subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java
Comment (mondokm) [reply subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java:1]: We do need something similar, but not XSTS-specific and much simpler as most of it was created for the int->enum literal traceability. which is not needed anymore.

Comment (RipplB) [reply subprojects/xsts/xsts/src/main/java/hu/bme/mit/theta/xsts/type/XstsCustomType.java:1]: Done with 5dec907

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java [hunk 2]
```diff
@@ -155,4 +141,9 @@ public GtExpr<FpType> Gt(Expr<FpType> leftOp, Expr<FpType> rightOp) {
     public GeqExpr<FpType> Geq(Expr<FpType> leftOp, Expr<FpType> rightOp) {
         return FpExprs.Geq(leftOp, rightOp);
     }
+
+    @Override
+    public DomainSize getDomainSize() {
+        return DomainSize.INFINITY;
```
Comment (as3810t) [inline subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java:147]: Technically, this is not true. We are storing the value on finite amount of bits.

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java [hunk 3]
```diff
@@ -175,6 +154,14 @@ private <I extends Type, E extends Type> Expr<?> createIndexValueArrayLitExpr(I
 
     ////////
 
+    private Expr<?> transform(final com.microsoft.z3.Expr term, final Model model,
+                              final List<Decl<?>> vars, final Type type) {
+        if (type instanceof EnumType enumType) {
+            return transformEnumLit(term, enumType);
+        }
```
Comment (as3810t) [inline subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java:161]: Can we refactor it to resemble how we handle arrays or function literals? `toFuncLitExpr`, `toArrayLitExpr`

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java [hunk 4]
```diff
@@ -494,8 +487,20 @@ private TriFunction<com.microsoft.z3.Expr, Model, List<Decl<?>>, Expr<?>> exprBi
...
+                op1 = transform(args[0], model, vars);
+            } catch (Exception e) {
+                op1 = transform(args[1], model, vars);
+                isInverse = true;
+            }
+            try {
+                op2 = transform(args[isInverse ? 0 : 1], model, vars);
+            } catch (Exception e) {
+                op2 = transform(args[isInverse ? 0 : 1], model, vars, op1.getType());
+            }
```
Comment (as3810t) [inline subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java:503]: I strongly recommend against this pattern, as it completely shatters the control flow. Moreover, catching `Exception` should not happen in code, as it also catches all of the validations we do. Why is this needed exactly? I am sure we can come up with a better solution.

File: subprojects/xta/xta/src/main/java/hu/bme/mit/theta/xta/utils/ClockType.java [hunk 5]
```diff
@@ -33,4 +34,8 @@ public String toString() {
         return "Clock";
     }
 
+    @Override
+    public DomainSize getDomainSize() {
+        return DomainSize.ONE;
```
Comment (as3810t) [inline subprojects/xta/xta/src/main/java/hu/bme/mit/theta/xta/utils/ClockType.java:39]: I am not sure if this is correct.

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		return of(operator.apply(left.finiteSize, right.finiteSize));
+	}
+
+	public boolean isInfinite() {
+		return equals(INFINITY);
+	}
+
+	public boolean isBiggerThan(long limit) {
+		return finiteSize.compareTo(BigInteger.valueOf(limit)) > 0;
+	}
```
Comment (as3810t) [inline subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java:56]: I think it will not work for `INFINITY` as intended.

**Review by as3810t — CHANGES_REQUESTED** [review]

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import com.google.common.base.Objects;
+
+import java.math.BigInteger;
+import java.util.function.BinaryOperator;
+
+public class DomainSize {
+
+	private final BigInteger finiteSize;
+
+	private DomainSize(BigInteger value) {
```
[the remark this answers is not in the document]
Comment (RipplB) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java:29]: Done in a315e11

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/bvtype/BvType.java [hunk 8]
```diff
@@ -223,4 +205,9 @@ public GeqExpr<BvType> Geq(Expr<BvType> leftOp, Expr<BvType> rightOp) {
             return BvExprs.UGeq(leftOp, rightOp);
         }
     }
+
+    @Override
+    public DomainSize getDomainSize() {
+        return DomainSize.of((long) Math.pow(2, size));
```
[the remark this answers is not in the document]
Comment (RipplB) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/bvtype/BvType.java]: Done in a315e11

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/enumtype/EnumLitExpr.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	private final EnumType type;
+	private final String value;
+
+	private EnumLitExpr(EnumType type, String value) {
+		this.type = type;
+		this.value = value;
+	}
+
+	public static EnumLitExpr of(EnumType type, String literalName) {
+		String value = literalName.contains(".") ? literalName.substring(literalName.indexOf(".") + 1) : literalName;
```
[the remark this answers is not in the document]
Comment (RipplB) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/enumtype/EnumLitExpr.java]: Done in a315e11

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java (hunk 2 above)
Comment (RipplB) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java]: Should be fixed with a315e11, please re-check

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java (hunk 6 above)
Comment (RipplB) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/DomainSize.java]: Done in a315e11

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java (hunk 3 above)
Comment (RipplB) [reply subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java]: These lines are removed in 0aa36d3

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java (hunk 4 above)
Comment (RipplB) [reply subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3TermTransformer.java]: The correct commit is 6fd805f

File: subprojects/xta/xta/src/main/java/hu/bme/mit/theta/xta/utils/ClockType.java (hunk 5 above)
Comment (as3810t) [reply subprojects/xta/xta/src/main/java/hu/bme/mit/theta/xta/utils/ClockType.java:39]: As I browsed through the code, I think the `ClockType` builds on `RatType`, so this should be `INFINITY`,

**Review by as3810t — COMMENTED** [review]
Thanks for the changes. I think this should integrate well with the existing code base.

**Comment by RipplB** [comment]
Development is continued in https://github.com/ftsrg/theta/pull/272
